test/utils.jl

```diff
@@ -0,0 +1,81 @@
+using NiSparseArrays:imul!
+
+"""
+    forwarddiff_mv_jacobian(A, B) -> AbstractArray J 
+forwarddiff_mv_jacobian calculates the jacobian with respect to parameter A and B based on multiplication in ForwardDiff
+"""
+function forwarddiff_mm_jacobian(A, B)
+    pA = params(A)
+    pB = params(B)
+    pC = vcat(pA, pB)
+    ForwardDiff.jacobian(pC) do pC
+        pA = pC[1:length(pA)]
+        pB = pC[length(pA)+1:end]
+        A2 = fitparams(A, pA)
+        B2 = fitparams(B, pB)
+        pC = params(A2 * B2)
+    end
+end
+
+"""
+    nilang_mv_jacobian(A, x::AbstractArray) -> AbstractArray J
+nilang_mv_jacobian calculates the jacobian with respect to parameter A and x based on mvloss in NiLang
+"""
+function nilang_mm_jacobian(A, B)
+    pB, pA = params(B), params(A)
+    C = zeros(eltype(B), size(A, 1), size(B, 2))
+    pC = params(C)
+    J = zeros(eltype(pB), length(pC), length(pB) + length(pA))
+    for j=1:length(pC)
+        _, _, _, _, gA, gB = NiLang.AD.Grad(mmloss)(Val(1), 0.0, j, C, A, B)
+        J[j,:] = vcat(grad(params(gA)), grad(params(gB)))
+    end
+    return J
+end
+
+"""
+    mvloss(l, j, y::AbstractArray{<:Real}, A, x) -> scalar l 
+mvloss defines loss on sparse-matrix & dense vector/matrix multiplication(Real Type)   
+"""
+@i function mmloss(l, j, C::AbstractArray{<:Real}, A, B)
+    imul!(C, A, B, 1.0, 1.0)
+    l += C[j]
+end
+
+"""
+    mvloss(l, j, y::AbstractArray{<:Complex}, A, x) -> scalar l 
+mvloss defines loss on sparse-matrix & dense vector/matrix multiplication(Complex Type)   
+"""
+@i function mmloss(l, j, C::AbstractArray{<:Complex}, A, B)
+    imul!(C, A, B, 1.0, 1.0)
+    if j%2 == 1
+        l += C[div(j-1,2)+1].re
+    else
+        l += C[div(j-1,2)+1].im
+    end
+end
+
+"""
+    helper function params and fitparams
+        function params wraps parameters in track of gradient when given array/matrix.
+            params(v::DenseArray)
+            params(A::SparseMatrixCSC)
+        function fitparams fits the original datastructure with parameters in track of gradient.
+            fitparams(v::DenseArray)
+            fitparams(A::SparseMatrixCSC)
+"""
+params(B::DenseArray{<:Real}) = B[:]
+fitparams(B::DenseArray{<:Real}, p) = reshape(p, size(B))
+params(B::DenseArray{<:Complex{T}}) where T = collect(reinterpret(T, B[:]))
+fitparams(B::DenseArray{<:T}, p::DenseArray{T2}) where {T<:Complex,T2} = reshape(collect(reinterpret(Complex{T2}, p)), size(B))
+params(A::AbstractSparseMatrix) = params(A.nzval)
+fitparams(A::AbstractSparseMatrix, p) = SparseMatrixCSC(A.m, A.n, A.colptr, A.rowval, fitparams(A.nzval, p))
+function params(xA::Adjoint{<:Any,<:AbstractSparseMatrix})
+    nA = copy(xA)
+    params(nA.nzval)
+end
+function fitparams(xA::Adjoint{<:Any,<:AbstractSparseMatrix}, p) 
+    nA = copy(xA)
+    fitparams(nA, p)
+end
+
```

Comment:
It seems adjoint is a lazy computation and xA has no field m, n,  colptr and nzval (but you could get these fields in xA.parent ) .  Furthermore,  construction of CSR format is not supported in Julia (I think ), so I couldn't  construct SparseCSC format directly with adjoint xA.

Reply:
This is sufficient:

```julia
params(xA::Ajoint) = params(xA.parent)
fitparams(xA::Adjoint, p) = fitparams(xA.parent, p)
```

Reply:
> ```julia
> params(xA::Adjoint) = params(xA.parent)
> fitparams(xA::Adjoint, p) = fitparams(xA.parent, p)
> ```

I test it on my local machine, and ForwardDiff would throw an exception:
``` julia
Testing Running tests...
jacobian on adjoint matrix multiplication: Error During Test at /Users/lijie/Fudan/git/NiSparseArrays.jl/test/jacobian.jl:18
  Got exception outside of a @test
  DimensionMismatch("")
  Stacktrace:
    [1] mul!(C::Vector{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 7}}, A::SparseMatrixCSC{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 7}, Int64}, B::Vector{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 7}}, α::Bool, β::Bool)
      @ SparseArrays /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/SparseArrays/src/linalg.jl:31
    [2] *(A::SparseMatrixCSC{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 7}, Int64}, x::Vector{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 7}})
      @ SparseArrays /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/SparseArrays/src/linalg.jl:51
    [3] (::var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}})(pC::Vector{ForwardDiff.Dual{ForwardDiff.Tag{var"#3#4"{Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, Vector{Float64}}, Float64}, Float64, 
```

Reply:
```julia
params(xA::Adjoint) = adjoint(params(xA.parent))
fitparams(xA::Adjoint, p) = adjoint(fitparams(xA.parent, p))
```

maybe?

Reply:
> ```julia
> params(xA::Adjoint) = adjoint(params(xA.parent))
> fitparams(xA::Adjoint, p) = adjoint(fitparams(xA.parent, p))
> ```

Thanks for johnny's advice, but  I'm afraid it would not work.  ajoint(params(x)) changes the shape of the input vector,  it means,  if you send a input vector with m length,  adjoint would change it into a matrix with m elements, it is not the expected input for the interface. params(xA. parent) is also not suitable since the order of value in adjoint object is not the same with the parent object.  

I also test it on my local machine,  and here is the error information 

``` julia
Testing Running tests...
jacobian on adjoint matrix multiplication: Error During Test at /Users/lijie/Fudan/git/NiSparseArrays.jl/test/jacobian.jl:18
  Got exception outside of a @test
  ArgumentError: number of columns of each array must match (got (11, 1))
  Stacktrace:
    [1] _typed_vcat(#unused#::Type{Float64}, A::Tuple{Adjoint{Float64, Vector{Float64}}, Vector{Float64}})
      @ Base ./abstractarray.jl:1553
    [2] typed_vcat
      @ ./abstractarray.jl:1567 [inlined]
    [3] vcat(::Adjoint{Float64, Vector{Float64}}, ::Vector{Float64})
      @ SparseArrays /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/SparseArrays/src/sparsevector.jl:1120
    [4] forwarddiff_mm_jacobian(A::Adjoint{Float64, SparseMatrixCSC{Float64, Int64}}, B::Vector{Float64})
      @ Main ~/Fudan/git/NiSparseArrays.jl/test/utils.jl:9 
```